From 8e08da6b766aa63235c646a6888faf93c0471bfa Mon Sep 17 00:00:00 2001
From: Cristian Bercaru <cristian.bercaru@freescale.com>
Date: Mon, 25 Nov 2013 21:17:31 +0000
Subject: [PATCH 432/987] dpaa_eth: proxy: treat error when starting proxy port
 fails

[Original patch taken from QorIQ-SDK-V1.6-SOURCE-20140619-yocto.iso]

The proxy interface start might fail because of various reasons. In some of
these cases, such as failing to start the mac device, the driver has to stop the
FMan ports.

Signed-off-by: Cristian Bercaru <cristian.bercaru@freescale.com>
Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Change-Id: I10e8d494760f9ec31efed9a0716b0306ce68959c
Reviewed-on: http://git.am.freescale.net:8181/6854
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Reviewed-by: Ruxandra Ioana Radulescu <ruxandra.radulescu@freescale.com>
Reviewed-by: Marian Cristian Rotariu <marian.rotariu@freescale.com>
Reviewed-by: Thomas Trefny <Tom.Trefny@freescale.com>
Reviewed-on: http://git.am.freescale.net:8181/7672
Reviewed-by: Madalin-Cristian Bucur <madalin.bucur@freescale.com>
Tested-by: Madalin-Cristian Bucur <madalin.bucur@freescale.com>
---
 drivers/net/ethernet/freescale/dpa/dpaa_eth_proxy.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/freescale/dpa/dpaa_eth_proxy.c b/drivers/net/ethernet/freescale/dpa/dpaa_eth_proxy.c
index 1c1a121..5b8af8d 100644
--- a/drivers/net/ethernet/freescale/dpa/dpaa_eth_proxy.c
+++ b/drivers/net/ethernet/freescale/dpa/dpaa_eth_proxy.c
@@ -218,10 +218,16 @@ int dpa_proxy_start(struct net_device *net_dev)
 		if (netif_msg_drv(priv))
 			netdev_err(net_dev, "mac_dev->start() = %d\n",
 					_errno);
-		return _errno;
+		goto port_enable_fail;
 	}
 
 	return _errno;
+
+port_enable_fail:
+	for_each_port_device(i, mac_dev->port_dev)
+		fm_port_disable(mac_dev->port_dev[i]);
+
+	return _errno;
 }
 
 int dpa_proxy_stop(struct proxy_device *proxy_dev, struct net_device *net_dev)
-- 
1.9.1

